tools/mason/MasonArgParse.chpl

```diff
@@ -0,0 +1,236 @@
+  // TODO: Verify no duplicate names, flags defined by dev
+  // TODO: Make sure we don't shadow Chapel flags
+  // TODO: Make sure we don't shadow config vars  
+  // TODO: Implement Help message and formatting
+  // TODO: Add bool flags
+  // TODO: Add int opts
+  // TODO: Add positional arguments
+  // TODO: Add public github issue when available
+
+  if chpl_warnUnstable then
+    compilerWarning("ArgumentParser is unstable.");
+
+// A generic argument parser error
+  class ArgumentError : Error {
+    var msg:string;
+    proc init(msg:string) {
+      this.msg = msg;
+    }
+    override proc message() {
+      return msg;
+    }
+  }
+  
+// indicates a result of argument parsing
+  class Argument {
+    var present: bool=false;
+    var values: list(string);     
+    
+    proc getValue(){     
+      return this.values.first();
+    }
+    iter getValues(){
+      for val in values {
+        yield val;
+      }      
+    }    
+    proc hasValue(){
+      return !this.values.isEmpty();
+    }
+  }
+
+  // stores the definition of an option
+  class Action {
+    var name:string;
+    var numOpts:int;
+    var opts:[0..numOpts-1] string;
+    var numArgs:range;
+
+    // TODO: Decouple the argument from the action
+    // maybe pass a list to fill by reference and have the argparser populate
+    // the argument instead?
+    // also need a bool by ref to indicate presence of arg or not
+    proc match(args:[?argsD]string, startPos:int, ref myArg:Argument) throws {
+      var high = 0;
+      
+      // TODO: Replace this high bound with something more reasonable
+      if !this.numArgs.hasHighBound() {
+        high = 10000000000;
+      } else {
+        high = this.numArgs.high;
+      }
+      writeErr("expecting between " + numArgs.low:string + " and "+high:string);
+      var matched = 0;
+      var pos = startPos;
+      var next = pos+1;
+      writeErr("starting at pos: " + pos:string);
+      while matched < high && next <= argsD.high && !args[next].startsWith("-"){
+        pos=next;
+        next+=1;
+        matched+=1;
+        myArg.values.append(args[pos]);
+        myArg.present=true;
+        writeErr("matched val: " + args[pos] + " at pos: " + pos:string);     
+      }
+      if matched < this.numArgs.low {
+        throw new ArgumentError("\\".join(opts) + " not enough values");
+      }
+      return next;
+    }
+ }
+
+  record argumentParser {
+    var result: map(string, shared Argument);
+    var actions: map(string, owned Action);
+    var options: map(string, string);
+
+    proc parseArgs(arguments:[?argsD]string) throws {
+      compilerAssert(argsD.rank==1, "parseArgs requires 1D array");
+      writeErr("start parsing args");   
+      var k = 0;
+      // identify optionIndices where opts start
+      var optionIndices : map(string, int);
+      const zArgsD = {0..argsD.size-1};
+      var zeroArgs = arguments.reindex(zArgsD);
+      var argsList = new list(zeroArgs);
```

Comment:
I don't think you need to reindex to create a list from an array. The list will always be 0-based.

Reply:
The original intent was to ensure alignment between the indices of the `arguments` array and the `argsList`, but upon further inspection there was a hidden bug here because the `argsList` is mutating while `arguments` holds constant, invaliding the alignment. I updated the code to lookup the index of the element in the list rather than assume the indices are aligned, eliminating the need for `zArgsD` and `zeroArgs.` Thanks for pointing this out!